File: app/model/sql/SqlEnums.java
package model.sql;

import model.Enums.Selectable;
import model.ExTag;
import scala.Option;
import scala.Some;


public abstract class SqlEnums {

    public enum SqlExerciseType implements Selectable<SqlExerciseType> {
        SELECT, CREATE, UPDATE, INSERT, DELETE;


        public static Option<SqlExerciseType> byString(String str) {
            try {
                return new Some<>(valueOf(str));
            } catch (Exception e) {
                // ==> None!
                return Option.apply(null);
            }
        }
    }

    public enum SqlExTag implements ExTag, Selectable<SqlExTag> {
        SQL_JOIN("J", "Join"),
        SQL_DOUBLE_JOIN("2J", "Zweifacher Join"),
        TRIPLE_JOIN("3J", "Dreifacher Join"),
        SQL_ORDER_BY("O", "Reihenfolge"),
        SQL_GROUP_BY("G", "Gruppierung"),
        SQL_FUNCTION("F", "Funktion"),
        SQL_ALIAS("A", "Alias"),
        SQL_LIMIT("L", "Limitierung"),
        SUBSELECT("S", "Zweites Select innerhalb");

        public final String buttonContent;
        public final String title;

        SqlExTag(String theButtonContent, String theTitle) {
            buttonContent = theButtonContent;
            title = theTitle;
        }

        @Override
        public String buttonContent() {
            return buttonContent;
        }

        @Override
        public String title() {
            return title;
        }

        public static Option<SqlExTag> byString(String str) {
            try {
                return new Some<>(valueOf(str));
            } catch (Exception e) {
                // ==> None!
                return Option.apply(null);
            }
        }
    }

}

File: app/model/tools/sql/SqlEnums.scala
package model.tools.sql

import enumeratum.{EnumEntry, PlayEnum}
import model.{ExPart, ExTag}

import scala.collection.immutable
import scala.collection.immutable.IndexedSeq


sealed abstract class SqlExPart(val partName: String, val urlName: String) extends EnumEntry with ExPart

object SqlExParts extends PlayEnum[SqlExPart] {

  val values: immutable.IndexedSeq[SqlExPart] = findValues

  case object SqlSingleExPart extends SqlExPart("Bearbeiten", "solve")

}


sealed abstract class SqlExTag(val buttonContent: String, val title: String) extends ExTag with EnumEntry


object SqlExTag extends PlayEnum[SqlExTag] {

  override def values: IndexedSeq[SqlExTag] = findValues

  // Types of queries?

  case object SQL_SELECT extends SqlExTag("S", "SELECT")

  case object SQL_UPDATE extends SqlExTag("U", "UPDATE")

  case object SQL_INSERT extends SqlExTag("I", "INSERT")

  case object SQL_DELETE extends SqlExTag("D", "DELETE")

  // Constructs used in queryy

  case object SQL_JOIN extends SqlExTag("J", "Join")

  case object SQL_DOUBLE_JOIN extends SqlExTag("2J", "Zweifacher Join")

  case object SQL_TRIPLE_JOIN extends SqlExTag("3J", "Dreifacher Join")

  case object SQL_ORDER_BY extends SqlExTag("O", "Reihenfolge")

  case object SQL_GROUP_BY extends SqlExTag("G", "Gruppierung")

  case object SQL_FUNCTION extends SqlExTag("F", "Funktion")

  case object SQL_ALIAS extends SqlExTag("A", "Alias")

  case object SQL_LIMIT extends SqlExTag("L", "Limitierung")

  case object SQL_SUBSELECT extends SqlExTag("S", "Zweites Select innerhalb")

}
